File: src/python/graph/1_netlist_to_graph.py
# file: 1_netlist_to_graph.py
# Convert netlist to NetworkX graph with nodes

import networkx as nx # Graph, write_gml
import sys            # argv

def main():
    input_file, output_file = sys.argv[1], sys.argv[2]
    G = nx.Graph()
    with open(input_file) as f:
        for line in f:
            if not line.strip() or ':' not in line: continue
            net, nodes = line.strip().split(':', 1)
            G.add_node(net, type='net')
            for n in nodes.split(','):
                if '_' in n:
                    comp, pin = n.rsplit('_', 1)
                    pin_node = f"{comp}_{pin}"
                    if not G.has_node(comp): G.add_node(comp, type='component')
                    if not G.has_node(pin_node): G.add_node(pin_node, type='pin')
                    G.add_edge(comp, pin_node)
                    G.add_edge(pin_node, net)
                else:
                    if not G.has_node(n): G.add_node(n, type='component')
                    G.add_edge(n, net)
    nx.write_gml(G, output_file)

if __name__ == "__main__":
    main()

File: src/typescript/graph/1_netlist_to_graph.ts
import fs from 'fs';
import Graph from 'graphology';

export function netlistToGraph(inputPath: string, outputPath: string): Graph {
    console.log(`[GRAPH] 1. Converting netlist to graph: ${path.basename(inputPath)}`);
    const G = new Graph();
    if (!fs.existsSync(inputPath)) {
        console.log(`[WARNING] Netlist file not found: ${inputPath}`);
        return G;
    }
    
    const content = fs.readFileSync(inputPath, 'utf8');
    const lines = content.split(/\r?\n/);
    console.log(`[GRAPH] Read ${lines.length} lines from netlist.`);

    for (const line of lines) {
        if (!line.trim() || !line.includes(':')) continue;
        const [net, nodeStr] = line.split(':');
        const nodes = nodeStr.split(',');

        if (!G.hasNode(net)) G.addNode(net, { type: 'net', label: net });

        for (const n of nodes) {
            if (n.includes('_')) {
                const lastUnderscore = n.lastIndexOf('_');
                const comp = n.substring(0, lastUnderscore);
                const pin = n.substring(lastUnderscore + 1);
                const pinNode = `${comp}_${pin}`;

                if (!G.hasNode(comp)) G.addNode(comp, { type: 'component', label: comp });
                if (!G.hasNode(pinNode)) G.addNode(pinNode, { type: 'pin', label: pinNode });

                if (!G.hasEdge(comp, pinNode)) G.addEdge(comp, pinNode);
                if (!G.hasEdge(pinNode, net)) G.addEdge(pinNode, net);
            } else {
                if (!G.hasNode(n)) G.addNode(n, { type: 'component', label: n });
                if (!G.hasEdge(n, net)) G.addEdge(n, net);
            }
        }
    }

    fs.writeFileSync(outputPath, JSON.stringify(G.export()));
    return G;
}
